### PR TITLE
[expo-modules-core] Add web implementation for eventemitter

### PR DESCRIPTION
**Pull request by aleqsio**
# Why

We need a web implementation to provide the same API for web modules that expo-modules-core provides for native modules.

# How

Added a web only `CoreModule` that is injected into `globalThis` and provides an `EventEmitter` class with a JS only event emitter implementation.

# Test Plan

Added unit tests.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-modules-core/src/web/CoreModule.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import { ExpoGlobal } from '..';
+import { EventEmitter } from '../ts-declarations/EventEmitter';
+import { NativeModule } from '../ts-declarations/NativeModule';
+import { SharedObject } from '../ts-declarations/SharedObject';
+import uuid from '../uuid';
+
+class WebEventEmitter<TEventsMap extends Record<never, never>> implements EventEmitter {
+  private listeners: Map<any, Set<Function>>;
```
Comment (tsapeta) [inline packages/expo-modules-core/src/web/CoreModule.ts:8]:
```suggestion
  private listeners: Map<keyof TEventsMap, Set<Function>> = new Map();
```

File: packages/expo-modules-core/src/web/CoreModule.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  constructor() {
+    this.listeners = new Map();
+  }
+
```
Comment (tsapeta) [inline packages/expo-modules-core/src/web/CoreModule.ts:13]:
```suggestion
```

File: packages/expo-modules-core/src/web/CoreModule.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
+import { ExpoGlobal } from '..';
+import { EventEmitter } from '../ts-declarations/EventEmitter';
+import { NativeModule } from '../ts-declarations/NativeModule';
+import { SharedObject } from '../ts-declarations/SharedObject';
```
Comment (tsapeta) [inline packages/expo-modules-core/src/web/CoreModule.ts:4]:
```suggestion
import type { ExpoGlobal } from '../ts-declarations/global';
import type { EventEmitter } from '../ts-declarations/EventEmitter';
import type { NativeModule } from '../ts-declarations/NativeModule';
import type { SharedObject } from '../ts-declarations/SharedObject';
```

File: packages/expo-modules-core/src/index.ts [hunk 4]
```diff
@@ -9,6 +9,7 @@ import { CodedError } from './errors/CodedError';
 import { UnavailabilityError } from './errors/UnavailabilityError';
 
 import './sweet/setUpErrorManager.fx';
+import './web/index';
```
Comment (tsapeta) [inline packages/expo-modules-core/src/index.ts:12]: We don't want web-specific code to be included in bundles for other platforms. I know there is `index.web.ts`, but will this work on other platforms? I guess Metro would say it's unable to resolve the file path.

Comment (aleqsio) [reply packages/expo-modules-core/src/index.ts:12]: I tested it out and it worked on iOS – was surprised too.
If this is confusing I can add an empty `index.ts` – wdyt?

File: packages/expo-modules-core/src/web/CoreModule.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
+import type { EventEmitter as EventEmitterType } from '../ts-declarations/EventEmitter';
+import type { NativeModule as NativeModuleType } from '../ts-declarations/NativeModule';
+import type { SharedObject as SharedObjectType } from '../ts-declarations/SharedObject';
+import uuid from '../uuid';
+
+class EventEmitter<TEventsMap extends Record<never, never>> implements EventEmitterType {
+  private listeners: Map<keyof TEventsMap, Set<Function>> = new Map();
```
Comment (tsapeta) [inline packages/expo-modules-core/src/web/CoreModule.ts:7]: Can we lazy-init this map?

**Review by tsapeta — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

**Pull request merged**